src/components/GraphiQL.js

```diff
@@ -815,79 +784,78 @@ export class GraphiQL extends React.Component {
-      }
-    }
+    //Check if one of the boxes are already empty, return the index of the first empty box; stays -1 if none are empty
+    let emptyIdx;
+    this.state.queryList.forEach((box, i) => {
+      emptyIdx = !box.query ? i : emptyIdx;
+    });
 
-    if (!!query && renderAndEmpty) {
-      let queryListCopy = [...this.state.queryList];
-      for (let i = 0; i < queryListCopy.length; i += 1) {
-        if (queryListCopy[i].render && queryListCopy[i].query === "") {
-          queryListCopy[i].query = query;
-          queryListCopy[i].checked = true;
-        }
-      }
-      this.setState({ queryList: queryListCopy });
-    }
-    // and would not enter into this if statement's code block
-    if (!renderAndEmpty) {
+    if (!!query && emptyIdx !== undefined)
+      this.setState((prevState, props) => {
+        prevState.queryList[emptyIdx].query = query;
+        return {
+          queryList: prevState.queryList
+        };
+      });
+
+    // if there are no empty boxes, add one
+    if (emptyIdx === undefined) {
+      // _editorQueryID is currently being used for History, planning to refactor this.
       this._editorQueryID = this.state.queryList.length;
-      const queriesNum = [...this.state.queryList];
-      queriesNum.push({
-        id: queriesNum.length,
-        render: true,
-        query,
-        checked: true,
-        operationName: undefined
+      this.setState((prevState, props) => {
+        // generate psuedo random id
+        let uniqid = Math.floor(Math.random() * Date.now());
+        prevState.queryList.push({
+          id: uniqid,
+          query,
+          checked: true,
+          operationName: undefined
+        });
+        return {
+          queryList: prevState.queryList
+        };
       });
-      this.setState({ queryList: queriesNum });
     }
   };
 
-  handleCheckQueryToRun = e => {
-    const queriesNum = [...this.state.queryList];
-    for (let i = 0; i < queriesNum.length; i += 1) {
-      if (queriesNum[i].id.toString() === e.target.id.substring(9)) {
-        if (e.target.checked === true) {
-          queriesNum[i].checked = true;
-        } else {
-          queriesNum[i].checked = false;
-        }
-      }
-    }
-    this.setState({ queryList: queriesNum });
-  };
+  handleCheckQueryToRun = (id, isChecked) => {
+    let idX;
+    prevState.queryList.forEach((box, i) => {
+      idX = box.id === id ? i : null;
+    });
 
-  handleDeleteQueryBox = e => {
-    // temp solution that doesn't allow you to delete the last code mirror editor (because it will throw error)
-    let renders = 0;
-    for (let i = 0; i < this.state.queryList.length; i += 1) {
-      if (this.state.queryList[i].render) {
-        renders += 1;
-      }
+    if (idX) {
+      setState((prevState, props) => {
+        // find the query with the same id as the button
+        prevState.queryList[idX].checked = isChecked;
+        return {
+          queryList: prevState
```

Comment:
queryList: prevState.queryList

Reply:
Once the id is found, perform operation

Reply:
use for Loop so that we can break out of it once the id is found and operation is performed